Source/NSAutoresizingMaskLayoutConstraint.m

```diff
@@ -0,0 +1,292 @@
+   version 2.1 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; if not, write to the Free
+   Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110 USA.
+*/
+#import "NSAutoresizingMaskLayoutConstraint.h"
+
+@implementation NSAutoresizingMaskLayoutConstraint
+
+NSUInteger minXAttribute = 32;
+NSUInteger minYAttribute = 33;
+NSUInteger maxXAttribute = 36;
+NSUInteger maxYAttribute = 37;
+
++ (NSArray *)constraintsWithAutoresizingMask:
+                 (NSAutoresizingMaskOptions) autoresizingMask
+                                     subitem: (NSView *) subItem
+                                       frame: (NSRect) frame
+                                   superitem: (NSView *) superItem
+                                      bounds: (NSRect) bounds
+{
+  if (autoresizingMask == NSViewNotSizable)
+    {
+      return [NSArray array];
+    }
+
+  NSAutoresizingMaskLayoutConstraint *xConstraint =
+      [self _xConstraintForAutoresizingMask: autoresizingMask
+                                    subitem: subItem
+                                      frame: frame
+                                  superitem: superItem
+                                     bounds: bounds];
+  NSAutoresizingMaskLayoutConstraint *yConstraint =
+      [self _yConstraintForAutoresizingMask: autoresizingMask
+                                    subitem: subItem
+                                      frame: frame
+                                  superitem: superItem
+                                     bounds: bounds];
+  NSAutoresizingMaskLayoutConstraint *widthConstraint =
+      [self _widthConstraintForAutoresizingMask: autoresizingMask
+                                        subitem: subItem
+                                          frame: frame
+                                      superitem: superItem
+                                         bounds: bounds];
+  NSAutoresizingMaskLayoutConstraint *heightConstraint =
+      [self _heightConstraintForAutoresizingMask: autoresizingMask
+                                         subitem: subItem
+                                           frame: frame
+                                       superitem: superItem
+                                          bounds: bounds];
+
+  NSMutableArray *constraints = [NSMutableArray arrayWithCapacity:4];
+  if ((autoresizingMask & NSViewMinXMargin)
+      && (autoresizingMask & NSViewMaxXMargin))
+    {
+      [constraints addObject: widthConstraint];
+      [constraints addObject: xConstraint];
+    }
+  else
+    {
+      [constraints addObject: xConstraint];
```

Comment:
Is this correct for all of the remaining three cases:

- `autoresizingMask` has a flag for min X, but not max X
- `autoresizingMask` has a flag for max X, but not min X
- `autoresizingMask` has a flag for neither min X nor max X

If yes, can you write a comment explaining why?

If no, can you address this?

Same applies for Y.

Reply:
I'm not exactly clear on what you are asking. The purpose of the factory method is to convert a NSAutoresizingMask into the equivalent layout constraints. In order for a view to be correctly constrained it must have a x, y, width and height constraint. If the mask doesn't have a min/max flag for a dimension (x or y) we are going to use the frame origin dimension for the position dimensional constraint.

Reply:
Here is the test suite I wrote, which may offer a deeper understanding of the behaviour. I created the tests first on MacOS prior to writing the implementation. https://github.com/BennyKJohnson/ConstraintBasedLayoutTests/blob/master/NSAutoresizingLayoutConstraintTests.m